### PR TITLE
[FLINK-5051] Backwards compatibility for serializers in backens 

### PR DESCRIPTION
**Pull request by StefanRRichter**
This PR sits on top of PR #2781 and introduces future backwards compatibility for state serializers and backends. We do so by providing version compatibility checking for TypeSerializer and making the serializers mandatory part of a keyed backend's meta data in checkpoints (so that we have everything required to reconstruct states in a self contained way). A serialization proxy is introduced for keyed state backend and operator state backend. Currently this serialization proxy covers the meta data, not yet the actual data. For most parts, the PR essentially moves functionality to a different place or makes formats more explicit.

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @aljoscha @StephanEwen

**Comment by StefanRRichter** [comment]
CC @aljoscha @kl0u @StephanEwen

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	}
+
+	@Override
+	public void write(DataOutputView out) throws IOException {
+		super.write(out);
+
+		keySerializerProxy.write(out);
+
+		out.writeShort(namedStateSerializationProxies.size());
+		Map<String, Integer> kVStateToId = new HashMap<>(namedStateSerializationProxies.size());
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java]: What is this Map used for?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 2]
```diff
@@ -1373,4 +1431,71 @@ private KeyGroupsStateHandle runSnapshot(RunnableFuture<KeyGroupsStateHandle> sn
 		}
 		return snapshotRunnableFuture.get();
 	}
+
+	private static final TypeSerializer<String> WRONG_VERSION_SERIALIZER = new TypeSerializer<String>() {
```
Comment (aljoscha) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Is it wrong version or wrong type, right now?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java (hunk 1 above)
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java]: Leftover from the previous code. Will remove it.

**Comment by aljoscha** [comment]
Thanks for changing! I just realised, however, that we already have tests in `StateBackendTestBase` that check for restoring from wrong serialisers so we don't actually need those.

I will also split the changes into 4 commits, you already split into FLINK-5282 and FLINK-5283, the later commits I will split into "[FLINK-5041] Savepoint Backwards Compatibility 1.1 -> 1.2" since these are essentially fixups for that and are not related to FLINK-5051. The last commit will contain the actual changes for FLINK-5051.

What do you think?

**Comment by StefanRRichter** [comment]
Yes, the commits you mentioned are actually meant as fixups for my previous PR. I think it would be good to separate this a bit if possible. Thanks! :-)

**Comment by aljoscha** [comment]
I merged this.

Thanks a lot for your work. 👍 Could you please close this PR.

**Pull request closed without merging**